src/array.rs

```diff
@@ -0,0 +1,402 @@
+//! Pixel arrays and slices.
+//!
+//! Pixel arrays and slices can be used to store pixel data as a tightly packed sequence of bytes.
+//! For color types with less then 8 bits per pixel multiple pixels will be stored in a single byte.
+//!
+//! Most applications won't use the types in this module directly and instead use higher level
+//! objects, like [framebuffers](crate::framebuffer) or [raw images](crate::image::ImageRaw).
+
+use core::marker::PhantomData;
+
+use crate::{
+    common::{ColorType, OutOfBoundsError},
+    pixelcolor::{
+        raw::{order::DataOrder, RawData},
+        StorablePixelColor,
+    },
+};
+
+/// Common functions for pixel slices and arrays.
+pub trait PixelData: ColorType
+where
+    Self::Color: StorablePixelColor,
+{
+    /// The data order.
+    type Order: DataOrder<<Self::Color as StorablePixelColor>::Raw>;
+
+    /// Returns the raw pixel data.
+    fn data(&self) -> &[u8];
+
+    /// Returns the length in pixels.
+    fn len(&self) -> usize {
+        let bpp = <Self::Color as StorablePixelColor>::Raw::BITS_PER_PIXEL;
+
+        if bpp >= 8 {
+            self.data().len() / (bpp / 8)
```

Comment:
Rounding up isn't necessary here, because BITS_PER_PIXEL is guaranteed to be a multiple of 8 if it is >= 8. But I've added the `bytes_to_pixels` function.